contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,503 @@
+* [Installation](#installation)
+* [Versioning](#versioning)
+* [References](#references)
+
+_Authors:_
+
+* @RenaudWasTaken - Renaud Gaubert &lt;rgaubert@NVIDIA.com&gt;
+* @jiayingz - Jiaying Zhang &lt;jiayingz@google.com&gt;
+
+# Motivation
+
+Kubernetes currently supports discovery of CPU and Memory primarily to a
+minimal extent. Very few devices are handled natively by Kubelet.
+
+It is not a sustainable solution to expect every hardware vendor to add their
+vendor specific code inside Kubernetes to make their devices usable.
+
+Instead, we want a solution for vendors to be able to advertise their resources
+to Kubelet and monitor them without writing custom Kubernetes code.
+We also want to provide a consistent and portable solution for users to
+consume hardware devices across k8s clusters.
+
+This document describes a vendor independant solution to:
+  * Discovering and representing external devices
+  * Making these devices available to the containers, using these devices,
+    scrubbing and securly sharing these devices.
+  * Health Check of these devices
+
+Because devices are vendor dependant and have their own sets of problems
+and mechanisms, the solution we describe is a plugin mechanism that may run
+in a container deployed through the DaemonSets mechanism.
+The targeted devices include GPUs, High-performance NICs, FPGAs, InfiniBand,
+Storage devices, and other similar computing resources that require vendor
+specific initialization and setup.
+
+The goal is for a user to be able to enable vendor devices (e.g: GPUs) through
+the following simple steps:
+  * `kubectl create -f http://vendor.com/device-plugin-daemonset.yaml`
+  * When launching `kubectl describe nodes`, the devices appear in the node
+    status as `extensions.kubernetes.io/vendor-device`. Note: naming
+    convention is discussed PR [#844](https://github.com/kubernetes/community/pull/844)
+
+# Use Cases
+
+ * I want to use a particular device type (GPU, InfiniBand, FPGA, etc.)
+   in my pod.
+ * I should be able to use that device without writing custom Kubernetes code.
+ * I want a consistent and portable solution to consume hardware devices
+   across k8s clusters.
+
+# Objectives
+
+1. Add support for vendor specific Devices in kubelet:
+    * Through an extension mechanism.
+    * Which allows discovery and health check of devices.
+    * Which allows hooking the runtime to make devices available in containers
+      and cleaning them up.
+2. Define a deployment mechanism for this new API.
+3. Define a versioning mechanism for this new API.
+
+# Non Objectives
+
+1. Handling heterogeneous nodes and topology related problems
+2. Collecting metrics is not part of this proposal. We will only solve
+   Health Check.
+
+# TLDR
+
+At their core, device plugins are simple gRPC servers that may run in a
+container deployed through the pod mechanism.
+
+These servers implement the gRPC interface defined later in this design
```

Comment:
1. API consistency across CSI and CNI (and CNI++) is a goal for v1.9 before this feature graduates to beta
2. Plugins register themselves with the Kubelet. This helps with API negotiation. 
3. Plugins can talk to kubelet via a well known socket.
4. Service names are presented in this proposal.

> What is the high-level back-and-forth for operations?

This is described in great detail in this proposal @thockin .

> I'd like that to be an explicit goal of this effort, and for CSI to make sure it leaves room to be consistent. Thoughts?

+1. I hope that's not a blocker for the upcoming alpha launch in v1.8.